File: sdk/java-sdk/src/main/scala/akka/platform/javasdk/client/ComponentMethodRef.scala
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package akka.platform.javasdk.client

import akka.NotUsed
import akka.annotation.DoNotInherit
import akka.platform.javasdk.DeferredCall
import akka.platform.javasdk.Metadata

import java.util.concurrent.CompletionStage

/**
 * No argument component call representation, not executed until invoked or by some mechanism using the deferred call
 * (like a timer executing it later for example)
 *
 * Not for user extension
 */
@DoNotInherit
trait ComponentMethodRef[R] {
  def withMetadata(metadata: Metadata): ComponentMethodRef[R]
  def deferred(): DeferredCall[NotUsed, R]
  def invokeAsync(): CompletionStage[R]
}

/**
 * One argument component call representation, not executed until invoked or by some mechanism using the deferred call
 * (like a timer executing it later for example)
 *
 * @tparam R
 *   the return type of the call
 * @tparam A1
 *   the argument type of the call
 *
 * Not for user extension
 */
@DoNotInherit
trait ComponentMethodRef1[A1, R] {
  def withMetadata(metadata: Metadata): ComponentMethodRef1[A1, R]
  def deferred(arg: A1): DeferredCall[A1, R]
  def invokeAsync(arg: A1): CompletionStage[R]
}

/**
 * No argument component call representation, not executed until invoked. Used for component methods that cannot be
 * deferred.
 *
 * Not for user extension
 */
@DoNotInherit
trait ComponentInvokeOnlyMethodRef[R] {
  def withMetadata(metadata: Metadata): ComponentMethodRef[R]
  def invokeAsync(): CompletionStage[R]
}

/**
 * One argument component call representation, not executed until invoked Used for component methods that cannot be
 * deferred.
 *
 * @tparam R
 *   the return type of the call
 * @tparam A1
 *   the argument type of the call
 *
 * Not for user extension
 */
@DoNotInherit
trait ComponentInvokeOnlyMethodRef1[A1, R] {
  def withMetadata(metadata: Metadata): ComponentMethodRef1[A1, R]
  def invokeAsync(arg: A1): CompletionStage[R]
}

File: sdk/java-sdk/src/main/java/akka/platform/javasdk/client/ComponentMethodRef.java
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package akka.platform.javasdk.client;

import akka.NotUsed;
import akka.annotation.DoNotInherit;
import akka.platform.javasdk.DeferredCall;
import akka.platform.javasdk.Metadata;

import java.util.concurrent.CompletionStage;

/**
 * Zero argument component call representation, not executed until invoked or by some mechanism
 * using the deferred call (like a timer executing it later for example)
 *
 * @param <R> The type of value returned by executing the call
 *     <p>Not for user extension or instantiation, returned by the SDK component client
 */
@DoNotInherit
public interface ComponentMethodRef<R> {
  ComponentMethodRef<R> withMetadata(Metadata metadata);

  DeferredCall<NotUsed, R> deferred();

  CompletionStage<R> invokeAsync();
}
